### PR TITLE
rgw/dbstore: Use DB transactions for I/O atomicity

### PR DESCRIPTION
**Pull request by soumyakoduri**
Use SQLite transactions for atomic read & write of the objects.

Signed-off-by: Soumya Koduri <skoduri@redhat.com>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
Looks very plausible.  I guess we need some correctness test?

File: src/rgw/store/dbstore/common/dbstore.h [hunk 1]
```diff
@@ -1221,6 +1228,11 @@ class DB {
     virtual int FreeDBOps(const DoutPrefixProvider *dpp) { return 0; }
     virtual int InitPrepareParams(const DoutPrefixProvider *dpp, DBOpPrepareParams &params) = 0;
 
+    virtual int beginTransaction(const DoutPrefixProvider *dpp) { return 0; }
+    virtual int endTransaction(const DoutPrefixProvider *dpp) { return 0; }
+    virtual int commitTransaction(const DoutPrefixProvider *dpp) { return 0; }
+    virtual int rollbackTransaction(const DoutPrefixProvider *dpp) { return 0; }
```
Comment (cbodley) [inline src/rgw/store/dbstore/common/dbstore.h:1234]: should these be pure virtual like the others? it doesn't seem like transaction support should be optional for derived classes

File: src/rgw/store/dbstore/common/dbstore.cc [hunk 2]
```diff
@@ -1294,14 +1294,22 @@ int DB::Object::Read::read(int64_t ofs, int64_t end, bufferlist& bl, const DoutP
   bufferlist read_bl;
   uint64_t max_chunk_size = store->get_max_chunk_size();
 
+  // begin transaction to avoid object modification while reading
+  // chunks
+  int r = store->beginTransaction(dpp);
```
Comment (cbodley) [inline src/rgw/store/dbstore/common/dbstore.cc:1299]: i'm confused about how these transactions could be global to the store. if the goal is to support several rgw operations at once, don't each of the operations need a separate transaction, and the ability to associate their database ops with that transaction?

File: src/rgw/store/dbstore/common/dbstore.h (hunk 1 above)
Comment (soumyakoduri) [reply src/rgw/store/dbstore/common/dbstore.h:1234]: got it.. thanks

File: src/rgw/store/dbstore/common/dbstore.cc (hunk 2 above)
Comment (mattbenjamin) [reply src/rgw/store/dbstore/common/dbstore.cc:1299]: These slides by Tom Lane seem to describe a system that works, in the context of a multi-threaded web+database application, the way we were elaborating in the prior discussion--there doesn't seem to be any mention of needing mutliple database handles matched to the different threads, but I might be wrong.  If that were the case, though, it shouldn't be necessary to keep a consistent thread handling the RGW operation--but rather, to associate the right handle with the operation, no (there'd be some other rules, too)?
https://www.postgresql.org/files/developer/concurrency.pdf

**Pull request closed without merging**